titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/workflow/step/MustBeRelocatedTaskCollectorStep.java

```diff
@@ -83,55 +80,61 @@ public MustBeRelocatedTaskCollectorStep(ReadOnlyAgentOperations agentOperations,
-                logger.info("Found task with no job record. Ignoring it: jobId={}, taskId={}", task.getJobId(), task.getId());
-                return;
-            }
-
-            Job<?> job = jobOpt.get();
-
-            if (hasDisruptionBudget(job) && isSelfManaged(job)) {
-                SelfManagedDisruptionBudgetPolicy selfManaged = (SelfManagedDisruptionBudgetPolicy) job.getJobDescriptor().getDisruptionBudget().getDisruptionBudgetPolicy();
-
-                TaskRelocationPlan relocationPlan = TaskRelocationPlan.newBuilder()
-                        .withTaskId(task.getId())
-                        .withReason(TaskRelocationReason.TaskMigration)
-                        .withReasonMessage("Agent instance scheduled to remove: instanceId=" + agent.getId())
-                        .withRelocationTime(now + selfManaged.getRelocationTimeMs())
-                        .build();
-
-                TaskRelocationPlan previous = lastResult.get(task.getId());
-                boolean keepPrevious = previous != null &&
-                        (areEqualExceptRelocationTime(previous, relocationPlan) || previous.getRelocationTime() < relocationPlan.getRelocationTime());
-
-                result.put(task.getId(), keepPrevious ? previous : relocationPlan);
-            }
+            agentOperations.findInstanceGroup(instance.getInstanceGroupId()).ifPresent(instanceGroup ->
+                    checkIfNeedsRelocationPlan(job, task, instanceGroup, instance).ifPresent(reason ->
+                            result.put(task.getId(), buildRelocationPlan(job, task, reason))
+                    ));
         });
 
         this.lastResult = result;
+        
+        return result;
+    }
 
+    private List<Triple<Job<?>, Task, AgentInstance>> findAllJobTaskAgentTriples() {
+        Map<String, AgentInstance> taskToInstanceMap = RelocationUtil.buildTasksToInstanceMap(agentOperations, jobOperations);
+
+        List<Triple<Job<?>, Task, AgentInstance>> result = new ArrayList<>();
+        jobOperations.getJobs().forEach(job -> {
+            jobOperations.getTasks(job.getId()).forEach(task -> {
+                TaskState taskState = task.getStatus().getState();
+                if (taskState != TaskState.Accepted && taskState != TaskState.KillInitiated && taskState != TaskState.Finished) {
+                    AgentInstance instance = taskToInstanceMap.get(task.getId());
+                    if (instance != null) {
+                        result.add(Triple.of(job, task, instance));
+                    }
+                }
+            });
+        });
         return result;
     }
 
-    private boolean isSelfManaged(Job<?> job) {
-        DisruptionBudgetPolicy disruptionBudgetPolicy = job.getJobDescriptor().getDisruptionBudget().getDisruptionBudgetPolicy();
-        return disruptionBudgetPolicy instanceof SelfManagedDisruptionBudgetPolicy;
+    private TaskRelocationPlan buildRelocationPlan(Job<?> job, Task task, String reason) {
```

Comment:
rename this as it is was not obvious this was only applicable for self managed relocations unless you go up the call stack.

Reply:
Renamed